src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -1018,6 +1039,51 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
 
         (qualifs.encode_to_bits(), self.tcx.arena.alloc(promoted_temps))
     }
+
+    /// Get the subset of `unchecked_promotion_candidates` that are eligible
+    /// for promotion.
+    // FIXME(eddyb) replace the old candidate gathering with this.
+    fn valid_promotion_candidates(&self) -> Vec<Candidate> {
```

Comment:
I believe this function is always called after const checking, so the `has_mut_interior` and `needs_drop` bitsets contain the qualifs at the end of the const body. I think this is correct because we only promote locals with exactly one definition. `NeedsDrop` is the only qualif that gets cleared when a local is moved out of, and obviously a local is never used once moved from, but this is subtle and should be documented.

Reply:
I think the right thing to do is to rely on your new infrastructure for the sanity check here, to compute qualifs.

Reply:
The non-dataflow, `TempPromotionResolver` lives in [this branch](https://github.com/ecstatic-morse/rust/tree/temp-only-qualif). I can open a draft PR if you'd rather.